OutboxSample/Application/DataAccess/IRepository.cs

```diff
@@ -0,0 +1,68 @@
+﻿namespace OutboxSample.Application.DataAccess;
+
+public interface IRepository
+{
+}
+
+public abstract record QueryResult
+{
+    public static Success<TData> OfSuccess<TData>(TData data, DataStore.ItemMetadata metadata) => new(data, metadata);
+
+    public sealed record Success<TData>(TData Data, DataStore.ItemMetadata Metadata) : QueryResult<TData>, Success;
+
+    public interface Success { }
+
+    public interface Failure { }
+}
+
+public abstract record QueryResult<TExpectedData> : QueryResult
+{
+    public static class OfFailure
+    {
+        public static Failure AlreadyExists(string message = "Already exists.") => new Failure.AlreadyExists(message);
+
+        public static Failure NotFound(string message = "Not found.") => new Failure.NotFound(message);
+
+        public static Failure ConcurrencyConflict(string message = "Concurrency conflict.") => new Failure.ConcurrencyConflict(message);
+    }
+
+    new public abstract record Failure(Failure.Description WhatHappened) : QueryResult<TExpectedData>, QueryResult.Failure
+    {
+        public sealed record AlreadyExists(string Message) : Failure(Message, ErrorCode.AlreadyExists);
+
+        public sealed record NotFound(string Message) : Failure(Message, ErrorCode.NotFound);
+
+        public sealed record ConcurrencyConflict(string Message) : Failure(Message, ErrorCode.ConcurrencyConflict);
+
+        protected Failure(string message, ErrorCode errorCode) : this(new Description(message, errorCode))
+        { }
+
+        public readonly record struct Description(string Text, ErrorCode ErrorCode);
+
+        public enum ErrorCode
+        {
+            Undefined = 0,
+            // Unexpected = 1,
+            NotFound = 2,
+            AlreadyExists = 3,
+            ConcurrencyConflict = 4,
+        }
+    }
+}
+
+public static class DataStore
+{
+    // NOTE: need to watch out for how much data it carries around as after some threshold it might make sense to conver it to ref-type record
+    public readonly record struct Item<TData>(TData Data, ItemMetadata Metadata);
+
+    public readonly record struct ItemMetadata(DateTime CreateAt, DateTime UpatedAt, uint Version)
```

Comment:
Do you anticipate that when storing `ItemMetadata`, these are additional columns in a database table? Would it be easier to have an interface and base class that define the properties of `ItemMetadata`, instead of a separate structure?

Reply:
Yes, those will be extra columns in a table. As for extra fields and interface - this is what I intentionally tried to avoid in this design. Mixing plumbing level (meta)data like that with domain-level data in out business entities is a bad design. Business-level entities should be represented by C# classes/structs that only carry data relevant for business use-cases. Otherwise we'd be coupling them to the 'plumbing'-level concerns. E.g. the `User` entity couldn't care less about when a DB row representing it's data was created(the `CreatedAt` field.